src/finn/custom_op/fpgadataflow/rtl/thresholding_rtl.py

```diff
@@ -0,0 +1,794 @@
+- output 0 is the output tensor, shape (..., NumChannels) - same as input
+- the '...' here can be any shape (representing groups of vectors)
+
+This module creates an RTL IP, HLS is not supported. See 'thresholding_batch'
+for a HLS equivalent.
+"""
+
+class Thresholding_rtl(Thresholding, RTLBackend):
+    """Class that corresponds to finn-rtllib 'thresholding' function."""
+
+    def __init__(self, onnx_node, **kwargs):
+        super().__init__(onnx_node, **kwargs)
+
+    def get_nodeattr_types(self):
+        my_attrs = {
+            # memory depth triggers for threshold storage
+            "depth_trigger_uram": ("i", False, 0),
+            "depth_trigger_bram": ("i", False, 0),
+            # enable uniform thres optimization
+            # doesn't actually do anything yet, only
+            # for resource estimations
+            "uniform_thres": ("i", False, 0, {0, 1}),
+            # enable deep pipelining for easier timing closure
+            # setting to 0 may save some FFs but otherwise leave on
+            "deep_pipeline": ("i", False, 1, {0, 1}),
+        }
+        my_attrs.update(Thresholding.get_nodeattr_types(self)) 
+        my_attrs.update(RTLBackend.get_nodeattr_types(self))
+        return my_attrs
+
+    def get_pe_mem_geometries(self):
+        pe = self.get_nodeattr("PE")
+        wdt = self.get_weight_datatype()
+        wdt_bits = wdt.bitwidth()
+        odt = self.get_output_datatype()
+        odt_bits = odt.bitwidth()
+        t_channels = self.get_nodeattr("NumChannels")
+        cf = t_channels / pe
+        is_uniform = self.get_nodeattr("uniform_thres")
+        if is_uniform:
+            ret = [(odt_bits - x, cf * (2**x)) for x in range(1, odt_bits)]
+        else:
+            ret = [(wdt_bits, (cf) * 2**x) for x in range(odt_bits)]
+        return ret
+
+    def get_memory_estimate(self):
+        res_dict = {}
+        depth_trigger_bram = self.get_nodeattr("depth_trigger_bram")
+        depth_trigger_uram = self.get_nodeattr("depth_trigger_uram")
+        pe = self.get_nodeattr("PE")
+        ret = self.get_pe_mem_geometries()
+        for mem_cfg in ret:
+            (width, depth) = mem_cfg
+            primitives = mem_primitives_versal
+            if depth_trigger_bram != 0 or depth_trigger_uram != 0:
+                if depth >= depth_trigger_bram and depth < depth_trigger_uram:
+                    primitives = {k: v for (k, v) in mem_primitives_versal.items() if "BRAM" in k}
+                elif depth >= depth_trigger_uram:
+                    primitives = {k: v for (k, v) in mem_primitives_versal.items() if "URAM" in k}
+            alts = get_memutil_alternatives(mem_cfg, primitives)
+            primary_alt = alts[0]
+            res_type = primary_alt[0].split("_")[0]
+            res_count, eff, waste = primary_alt[1]
+            res_dict[res_type] = res_dict.get(res_type, 0) + pe * res_count
+        return res_dict
+
+    def calc_tmem(self):
```

Comment:
calc_tmem(self) is identical between HLS and RTL variant and thus can be moved into HW abstraction layer. It can then be deleted from the HLS/RTL variants because it gets inherited from `Thresholding`.